File: src/test/java/com/testainers/DelayResourceTest.java
package com.testainers;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.Method;
import org.junit.jupiter.api.Test;

import static org.hamcrest.Matchers.containsStringIgnoringCase;
import static org.hamcrest.Matchers.is;

/**
 * @author Eduardo Folly
 */
@QuarkusTest
public class DelayResourceTest extends BaseResourceTest {

    @Test
    public void delayEmpty() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay")
                    .then()
                    .statusCode(404)
                    .statusLine(containsStringIgnoringCase("Not Found"));
        }
    }

    @Test
    public void delayString() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/a")
                    .then()
                    .statusCode(404)
                    .statusLine(containsStringIgnoringCase("Not Found"));
        }
    }

    @Test
    public void delayDouble() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/1.8")
                    .then()
                    .statusCode(404)
                    .statusLine(containsStringIgnoringCase("Not Found"));
        }
    }

    @Test
    public void delayNegative() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/-1")
                    .then()
                    .statusCode(400)
                    .body("body", is("Invalid delay: -1"),
                          bodyMatchers(method));
        }
    }

    @Test
    public void delay0() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/0")
                    .then()
                    .statusCode(200)
                    .body("body", is("Slept for 0 seconds."),
                          bodyMatchers(method));
        }
    }

    @Test
    public void delay1() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/1")
                    .then()
                    .statusCode(200)
                    .body("body", is("Slept for 1 seconds."),
                          bodyMatchers(method));
        }
    }

    @Test
    public void delay10() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/10")
                    .then()
                    .statusCode(200)
                    .body("body", is("Slept for 10 seconds."),
                          bodyMatchers(method));
        }
    }

    @Test
    public void delay11() {
        for (Method method : METHODS) {
            json(method)
                    .request(method, "/delay/11")
                    .then()
                    .statusCode(400)
                    .body("body", is("Invalid delay: 11"),
                          bodyMatchers(method));
        }
    }

}

File: src/test/kotlin/com/testainers/DelayResourceTest.kt
package com.testainers

import io.quarkus.test.junit.QuarkusTest
import io.restassured.http.Method
import org.hamcrest.Matchers.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.*

/**
 * @author Eduardo Folly
 */
@QuarkusTest
class DelayResourceTest : BaseResourceTest() {
    companion object : BaseResourceTest() {
        @JvmStatic
        fun invalidDelay(): List<Arguments> = argumentGenerator(listOf(-1, 11))

        @JvmStatic
        fun validDelay(): List<Arguments> = argumentGenerator(listOf(0, 1))
    }

    @ParameterizedTest
    @MethodSource("onlyMethods")
    fun delayEmpty(method: Method) {
        json(method)
            .request(method, "/delay")
            .then()
            .statusCode(404)
            .statusLine(containsStringIgnoringCase("Not Found"))
    }

    @ParameterizedTest
    @MethodSource("notFoundStatus")
    fun notFound(
        method: Method,
        delay: String?,
    ) {
        json(method)
            .request(method, "/delay/$delay")
            .then()
            .statusCode(404)
            .statusLine(containsStringIgnoringCase("Not Found"))
    }

    @ParameterizedTest
    @MethodSource("invalidDelay")
    fun invalid(
        method: Method,
        delay: Int,
    ) {
        json(method)
            .request(method, "/delay/$delay")
            .then()
            .statusCode(400)
            .apply {
                if (method != Method.HEAD) {
                    body(
                        "body",
                        equalTo("Invalid delay: $delay"),
                        *bodyMatchers(method),
                    )
                }
            }
    }

    @ParameterizedTest
    @MethodSource("validDelay")
    fun valid(
        method: Method,
        delay: Int,
    ) {
        json(method)
            .request(method, "/delay/$delay")
            .then()
            .statusCode(200)
            .apply {
                if (method != Method.HEAD) {
                    body(
                        "body",
                        equalTo("Slept for $delay seconds."),
                        *bodyMatchers(method),
                    )
                }
            }
    }
}
